src/main/java/uk/ac/ebi/biostudies/controller/FileDownload.java

```diff
@@ -44,44 +52,55 @@ public class FileDownload {
         String dlType = request.getParameter("type");
         String os = request.getParameter("os");
-        if(os==null || os.isEmpty())
-            os="unix";
+        if (os == null || os.isEmpty())
+            os = "unix";
         String fileExtension = "sh";
         fileExtension = getFileExtension(os);
         List<String> fileNames = new ArrayList<>();
-        Document submissionDoc = getFilePaths(request,response, fileNames);
+        Document submissionDoc = getFilePaths(request, response, fileNames);
         String relativeBaseDir = submissionDoc.get(Constants.Fields.RELATIVE_PATH);
         String accession = submissionDoc.get(Constants.Fields.ACCESSION);
         String storageModeString = submissionDoc.get(Constants.Fields.STORAGE_MODE);
         Constants.File.StorageMode storageMode = Constants.File.StorageMode.valueOf(StringUtils.isEmpty(storageModeString) ? "NFS" : storageModeString);
         dlType = dlType.replaceFirst("/", "");
         String[] files = request.getParameterMap().get("files");
-        if(dlType.equalsIgnoreCase("zip"))
+        if (dlType.equalsIgnoreCase("zip"))
             zipDownloadService.sendZip(request, response, files, storageMode);
-        else if(dlType.equalsIgnoreCase("ftp") || dlType.equalsIgnoreCase("aspera")){
-            response.setContentType("application/txt");
-            response.addHeader("Content-Disposition", "attachment; filename="+accession+"-" + os+"-"+dlType+"."+fileExtension);
+        else if (dlType.equalsIgnoreCase("ftp") || dlType.equalsIgnoreCase("aspera")) {
+            List<String> newFileNames = new ArrayList<>();
+            if (storageMode == Constants.File.StorageMode.FIRE) {
+                newFileNames = Arrays.stream(files).map(name -> {
+                    try {
+                        return jsEngine.eval(String.format("unescape(encodeURIComponent('%s'))", name)).toString().replace("#", "%23").replace("+", "%2B").replace("=", "%3D").replace("@", "%40").replace("$", "%24");
+                    } catch (ScriptException e) {
+                        LOGGER.error(name + " problem in unescapeing", e);
+                    }
+                    return "";
+                }).collect(Collectors.toList());
+            }
+            response.setContentType("application/txt; charset=UTF-8");
+            response.addHeader("Content-Disposition", "attachment; filename=" + accession + "-" + os + "-" + dlType + "." + fileExtension);
             response.addHeader("Cache-Control", "no-cache");
-            response.getOutputStream().print(batchDownloadScriptBuilder.fillTemplate(dlType, fileNames, relativeBaseDir, os, storageMode));
-            response.getOutputStream().close();
+            response.getWriter().print(batchDownloadScriptBuilder.fillTemplate(dlType, storageMode == Constants.File.StorageMode.FIRE ? newFileNames : fileNames, relativeBaseDir, os, storageMode));
```

Comment:
No check would be needed if we just use the newFileNames and do the replacement in a function (as specified above)